Ble connection tests using nrf52_bsim

aescolar opened the pull request:
This PR adds 2 BLE test cases which use the nrf52_bsim board (and BabbleSim).
The testcases themselves are based on the central_hr sample application.
The testcase is considered passed, if during the first 5 seconds after boot, it manages to find and connect to a peripheral and a notification is received from it.
One will encrypt the link, the other won't

A couple of bare-bones, scripts are added to compile and run the tests.

aescolar committed:
tests: Add new BabbleSim based BT test

Test which depends on the nrf52_bsim board.
It is based on the central_hr sample application.
The testcase is considered passed, if during the first 5 seconds
after boot, it manages to find and connect to a
peripheral and a notification is received from it.
Otherwise, the testcase fails.
Note that the executable return code will reflect the status of the
test:
0: Testcase passed
1: Testcase was stopped while in progress
2: Testcase failed
anything else: A failure not from the testcase itself

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
tests bsim_bt: Add script to execute the basic connection test

A script which will execute:
* the peripheral sample
* the bsim_test_app (the actual self-testing application)
* the BabbleSim phy

This script will return
* 0 if the test passes
* something else otherwise

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
tests bsim_bt: Add scripts to compile and run tests

The compile.sh script compiles the neccessary applications to run
all testcases.

The run_parallel.sh script runs all avaliable tests scripts
(e.g.
tests/bluetoothbsim_bt/bsim_test_app/tests_scripts/Basic_con.sh )
and reports which of them pass or fail.

Note that the run_parallel script will run the test in parallel
only if it can find GNU parallel (which is missing in CI),
otherwise it just runs them in serial

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
gitignore: Ignore default bsimt_bt test output folder

Similarly to sanitycheck the bsim_bt tests have their own
work folder where compilation output and coverage results are
kept.
Ignore the default folder

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
tests: Add new bsim based encrypted BLE link test

Test which depends on the nrf52_bsim board.
It is based on the basic connection bsim test,
with the same pass/fail critaria.
The only difference being that the link will be encrypted
therefore exercising that part of the BLE stack.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
In the last commit, I have just added a test based on the previous, but with the link encrypted.
It bumps the coverage quite a lot.

jhedberg commented on tests/bsim_bt/bsim_test_app/src/test_connect1.c:
```diff
@@ -0,0 +1,368 @@
+}
+
+static const bst_test_instance_t test_unencrypted[] = {
+	{"connect",        /* Test Name */
```
Doesn't the coding style recommend using explicit struct member initializers like:
```
{
        .test_id = "connect",
        .test_descr = "Basic connection test.",
        ...
}
```

This is IMO more readable, you don't need the intermediate NULL entries (just leave something out if you don't use it) and it's also more robust since now you don't depend on the ordering of variables inside the struct.

Btw, why is this struct a typedef? AFAIK only types where the API user is not expected to directly access the struct internals should be typedefs (i.e. the meaning a typedef conveys is that the type is considered opaque)

aescolar replied:
Thanks Johan. I agree. Fixed.

aescolar commented:
And by the way, if you feel like generating the coverage report locally,
after running the tests:
```
source zephyr-env.sh
export BSIM_OUT_PATH=/opt/bsim/
export BSIM_COMPONENTS_PATH=${BSIM_OUT_PATH}/components/
WORK_DIR=${ZEPHYR_BASE}/bsim_bt_out tests/bsim_bt/compile.sh
RESULTS_FILE=/dev/null SEARCH_PATH=tests/bsim_bt/bsim_test_app/tests_scripts tests/bsim_bt/run_parallel.sh
```
You can generate the report:
```
rm lcov.*info ; true
lcov_branch_coverage=1  2>&1 | grep -v "pref"
lcov --capture --directory bsim_bt_out/ --output-file lcov.pre.info -q --rc lcov_branch_coverage=1  2>&1 | grep -v "pref"
lcov -q --remove lcov.pre.info mylib.c --remove lcov.pre.info *generated* -o lcov.info -rc lcov_branch_coverage=1
genhtml lcov.info --output-directory lcov_html -q --ignore-errors source  --branch-coverage --highlight --legend 
```
And open it:
```
firefox-esr lcov_html/index.html &> /dev/null
```

aescolar commented:
recheck

aescolar commented:
recheck

aescolar commented:
@carlescufi : As discussed today, please remeber to take a look at this PR

aescolar commented on tests/bsim_bt/run_parallel.sh:
```diff
@@ -0,0 +1,95 @@
+
+set -u
+
+TODAY=$(date '+%Y/%m/%d %H:%M:%S')
```
self comment: not used anymore, remove

nashif commented on tests/bsim_bt/bsim_test_app/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+endif()
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(NONE)
```
give it some name, not NONE

aescolar replied:
Thanks @nashif , your comments are addressed.

nashif reviewed (changes requested):
Please move those tests either under tests/bluetooth or under tests/boards/..

nashif commented on .shippable.yml:
```diff
@@ -77,6 +78,11 @@ build:
             fi;
             rm -f modified_tests.args modified_boards.args;
           fi;
+      - >
```
you need to make the changes in scripts/ci/run_ci.sh now.

aescolar replied:
Yep. I'm out of office today, I will (hopefully) align it tomorrow

aescolar committed:
ci: Run bsim_bt tests and collect coverage results

Compile and run the tests avaliable in bsim_bt
and collect the coverage results into the coverage report.

Also, detect if bsim's component folder already contains the
nRF52 HW models, and if it does instead of trying to fetch
them again (which will fail) check that the right versio is
present. This should ease testing locally.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

carlescufi reviewed (approved):
I don't want to block this since it's already been idle for a while.
Obviously this is not future-proof or the way to go when it comes to incorporating additional functionality to our test framework (currently `sanitycheck` mostly), but I do believe it's better to have this in and testing BLE on CI than not having anything at all. So while we discuss the bigger, forward-looking plans for testing on Zephyr I believe we should merge this as-is now.
I hope the test working group can discuss the requirements so that we try to standardize on a single test framework in the future.

jhedberg reviewed (approved):
I'm along the same thoughts as @carlescufi 	just outlined

aescolar commented on scripts/ci/run_ci.sh:
```diff
@@ -107,8 +126,17 @@ function get_tests_to_run() {
+	build_btsim
+
+	# Run BLE tests in simulator on the 1st CI instance:
+	if [ "$MATRIX_BUILD" = "1" ]; then
```
I used MATRIX_BUILD from the .shippable environment.. but yep better to use MATRIX. Will change right away

aescolar replied:
Should I assume you do not want to use any variable in the script from the environment? (but to pass them as args to it?)

nashif replied:
yes, the idea to be able to run the script outside of CI for testing and reproducing issues

aescolar commented:
This is the only delta in the last push
```diff
diff --git a/scripts/ci/run_ci.sh b/scripts/ci/run_ci.sh
index eeb4879d65..f456f521b9 100755
--- a/scripts/ci/run_ci.sh
+++ b/scripts/ci/run_ci.sh
@@ -131,7 +131,7 @@ if [ ! -z "${BSIM_OUT_PATH}" ]; then
        build_btsim
 
        # Run BLE tests in simulator on the 1st CI instance:
-       if [ "$MATRIX_BUILD" = "1" ]; then
+       if [ "$MATRIX" = "1" ]; then
                run_bsim_bt_tests
        fi
 else
```
